src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocStyleCheck.java

```diff
@@ -387,30 +390,270 @@ public int[] getAcceptableTokens() {
     }
 
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
     @Override
     public void visitToken(DetailAST ast) {
         if (shouldCheck(ast)) {
-            final FileContents contents = getFileContents();
             // Need to start searching for the comment before the annotations
             // that may exist. Even if annotations are not defined on the
             // package, the ANNOTATIONS AST is defined.
-            final TextBlock textBlock =
-                contents.getJavadocBefore(ast.getFirstChild().getLineNo());
+            final TextBlock textBlock;
+            final boolean isPackageToken = ast.getType() == TokenTypes.PACKAGE_DEF;
+
+            if (isPackageToken) {
+                textBlock = getJavaDocForPackage(ast);
+            }
+            else {
+                textBlock = getJavaDoc(ast);
+            }
 
             checkComment(ast, textBlock);
         }
     }
 
+    /**
+     * Get The Javadoc For Package Token.
+     *
+     * @param ast as First AST Token
+     * @return TextBlock of Javadoc Comment
+     */
+    private TextBlock getJavaDocForPackage(DetailAST ast) {
+        DetailAST temp = ast;
+        String comment = null;
+        final int commentDetailsSize = 4;
+        int[] commentDetails = new int[commentDetailsSize];
+
+        while (temp.getPreviousSibling() != null) {
+            temp = temp.getPreviousSibling();
+        }
+
+        while (true) {
+            DetailAST toVisit = temp.getFirstChild();
+
+            if (temp.getType() == TokenTypes.BLOCK_COMMENT_BEGIN
+                    && JavadocUtil.isJavadocComment(temp)) {
+                final DetailAST end = temp.getFirstChild().getNextSibling();
+                comment = JavadocUtil.getJavadocCommentContent(temp);
+                commentDetails = commentLineDetails(temp, end);
+            }
+
+            if (temp.getType() == TokenTypes.PACKAGE_DEF) {
+                break;
+            }
+
+            while (toVisit == null) {
+                toVisit = temp.getNextSibling();
+                temp = temp.getParent();
+            }
+
+            temp = toVisit;
+        }
+
+        TextBlock textBlock = generateTextBlock(comment, commentDetails);
+
+        if (textBlock == null) {
+            textBlock = getJavaDoc(ast);
+        }
+
+        return textBlock;
+    }
+
+    /**
+     * Generate The TextBlock From Given Comment.
+     *
+     * @param comment as Javadoc Comment String
+     * @param commentDetails as Line Number and Column Number Details
+     *        commentDetails[0] as Start Line Number
+     *        commentDetails[1] as Start Column Number
+     *        commentDetails[2] as End Line Number
+     *        commentDetails[3] as End Column Number
+     * @return TextBlock of Javadoc comment
+     */
+    public TextBlock generateTextBlock(String comment, int... commentDetails) {
```

Comment:
If this is not avoidable, it is better to consider to upgrade Check to be javadoc check
https://checkstyle.org/writingjavadocchecks.html